### PR TITLE
Add dimension check to gridrec

### LINKED ISSUE #604: Missing dimension check in gridrec
**Describe the problem**
The `gridrec()` function in `grdirec.c` receives arrays and sizes from `c_gridrec()` in `grdirec.py`. It assumes that `dt` is the length of `theta` but this is not checked. If the length of `theta` is less than the t dimension of `tomo` then `set_trig_tables()` will read passed the end of the array.

**To Reproduce (if applicable)**

```python
import tomopy

obj = tomopy.shepp2d()
ang = tomopy.angles(180)
ang_bad = tomopy.angles(10)
sim = tomopy.project(obj, ang)

rec1 = tomopy.recon(sim, ang, algorithm='gridrec')
rec2 = tomopy.recon(sim, ang_bad, algorithm='gridrec')

print(rec1.mean(), rec2.mean())
```
Here the second call to `tomopy.recon()` will read values from beyond the `theta` array. The output for `rec2` will change run to run.

It can also be seen by putting a printf in the loop in `set_trig_tables()`.

**Expected behavior (if applicable)**
I'd expect an error message if the shapes do not match.

**Platform Information:**
 - OS: Ubuntu and WIndows
 - Python Version 3.7 and 3.9
 - TomoPy Version 1.12.2 and master

### PR DESCRIPTION
**Pull request by samtygier-stfc**
Check that theta array matches the `dt` that is being passed to the c function.

Closes #604

### CODE REVIEW

**Comment by samtygier-stfc** [comment]
Should I add the same for the other c wrappers?

**Review by carterbox — CHANGES_REQUESTED** [review]
Thank you for opening this PR to squash this bug.

This check should be located in `tomopy.recon.algorithm.recon()` near the top of the function. Add something like:

```python
    if tomo.shape[1] != len(theta):
        msg = 'There must be one angle for every projection.'
        raise ValueError(msg)
```

**Comment by samtygier-stfc** [comment]
My reason for putting it near the c call was to make it certain when considering the c code that the arrays and sizes passed in are compatible.

But I am happy to move it into `recon()` if you prefer.

**Comment by carterbox** [comment]
I prefer checking in `recon()` because it provides the best coverage, and the util.extern module is non-public, so it shouldn't be called directly.

**Comment by samtygier-stfc** [comment]
Ok, moved it into `recon()`. I think we can assume here that it is 3D, otherwise `_get_algorithm_kwargs` gives an error.

I'd like to also add a check for `center`.

**Review by carterbox — CHANGES_REQUESTED** [review]

File: source/tomopy/recon/algorithm.py [hunk 1]
```diff
@@ -310,6 +314,11 @@ def recon(tomo,
 
     # Generate args for the algorithm.
     center_arr = get_center(tomo.shape, center)
+
+    if tomo.shape[0] != len(center_arr):
+        msg = 'There must be one center for every slice.'
+        raise ValueError(msg)
+
```
[the remark this answers is not in the document]
Comment (samtygier-stfc) [reply source/tomopy/recon/algorithm.py:321]: It is needed if center is and array or list with a size different from 1, then its shape is preserved. So in:

```
obj = tomopy.shepp3d()
ang = tomopy.angles(180)
c = [5,5,5]
rec = tomopy.recon(sim, ang, center=c, algorithm="gridrec")
```
the wrong sized array is passed to `c_gridrec` and the c function uses values beyond its end.

**Review by carterbox — APPROVED** [review]

**Comment by carterbox** [comment]
Thanks for your contribution!

**Pull request merged**